hail/python/hail/expr/functions.py

```diff
@@ -3384,9 +3384,45 @@ def zip(*arrays, fill_missing: bool = False) -> ArrayExpression:
                           aggregations)
 
 
+@typecheck(a=expr_array(), start=expr_int32, index_first=bool)
+def enumerate(a, start=0, *, index_first=True):
+    """Returns an array of (index, element) tuples.
+
+    Examples
+    --------
+
+    >>> hl.eval(hl.enumerate(['A', 'B', 'C']))
+    [(0, 'A'), (1, 'B'), (2, 'C')]
+
+    >>> hl.eval(hl.enumerate(['A', 'B', 'C'], start=3))
+    [(3, 'A'), (4, 'B'), (5, 'C')]
+
+    >>> hl.eval(hl.enumerate(['A', 'B', 'C'], index_first=False))
+    [('A', 0), ('B', 1), ('C', 2)]
+
+
+    Parameters
+    ----------
+    a : :class:`.ArrayExpression`
+    start : :class:`.Int32Expression`
```

Comment:
I would say something like "the starting 'index'", index can be first or second.

Reply:
So the reason I worded it the way I did was because I was afraid to give the implication that "index" meant "indexing into `a` (i.e., specifying start = 2 would mean to start at position 2 of `a`)".

Reply:
Let's make our language consistent with Python's `enumerate` method, or if you feel strongly that this isn't the right direction, make our language internally consistent. We use "index" throughout, and "label" does not indicate that the number included in the tuple is ordinal. 

Here is how Python's enumerate is described elsewhere:

"The __next__() method of the iterator returned by enumerate() returns a tuple containing a count (from start which defaults to 0) and the values obtained from iterating over iterable."
  - https://docs.python.org/3/library/functions.html#enumerate

"And there is more! enumerate also accepts an optional argument that allows us to specify the starting index of the counter."
  - https://book.pythontips.com/en/latest/enumerate.html 

"Start: the index value from which the counter is 
              to be started, by default it is 0 "
  - https://www.geeksforgeeks.org/enumerate-in-python/


A few of our uses of "index":
"Returns an array of (index, element) tuples."
"Array of (index, element) or (element, index) tuples."
"The number the first element of the array is labeled with.
     index_first: :obj:`bool` ...  If ``True``, the index is the first value of the element tuples"

Reply:
You've convinced me. I'll go with language that includes "index", that seems to be the general understanding. I like "The index value from which the counter is started", I think that's clear.